### ISSUE TITLE
False positive in cmp_owned lint

### ISSUE THREAD

**Issue by tesuji** [rust-lang/rust-clippy#4909]
Consider this (crappy code)
```rust
macro_rules! foo {
    ($($i:ident: $ty:ty, $def:expr);+ $(;)*) => {
        pub fn bar(key: &str) -> bool {
            match key {
                $(
                    stringify!($i) => {
                        let res = <$ty>::default() == $def;
                        let _i: $ty = $def;
                        return res;
                    }
                )+
                _ => false,
            }
        }
    };
}

foo! {
    max_width: usize, 100;
    required_version: String, env!("HOME").to_string();
}
```

Clippy outputs this warnings:
```rust
warning: this creates an owned instance just for comparison
  --> src/lib.rs:20:31
   |
20 |     required_version: String, env!("HOME").to_string();
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `env!("HOME")`
   |
   = note: `#[warn(clippy::cmp_owned)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cmp_owned
```

Which is wrong.

---

`cargo clippy --version`: clippy 0.0.212 (69f99e7 2019-12-14)

**Comment by mikerite** [comment]
This is incorrect because the instance is being assigned in this line `let _i: $ty = $def;`, right?

**Comment by lzutao** [comment]
Yeah.

**Comment by tesuji** [comment]
The example code in OP was gotten from rustfmt `create_config!`. But since then, there are no real practical usage
of the same code pattern. Closing as not planned.
